packages/gatsby/src/recipes/providers/gatsby/plugin.js

```diff
@@ -0,0 +1,272 @@
+  if (property.type !== `Identifier` || property.name !== `exports`)
+    return false
+
+  return true
+}
+
+const getValueFromLiteral = node => {
+  if (node.type === `StringLiteral`) {
+    return node.value
+  }
+
+  if (node.type === `TemplateLiteral`) {
+    return node.quasis[0].value.raw
+  }
+
+  return null
+}
+
+const getNameForPlugin = node => {
+  if (node.type === `StringLiteral` || node.type === `TemplateLiteral`) {
+    return getValueFromLiteral(node)
+  }
+
+  if (node.type === `ObjectExpression`) {
+    const resolve = node.properties.find(p => p.key.name === `resolve`)
+    return resolve ? getValueFromLiteral(resolve.value) : null
+  }
+
+  return null
+}
+
+const addPluginToConfig = (src, pluginName) => {
+  const addPlugins = new BabelPluginAddPluginsToGatsbyConfig({
+    pluginOrThemeName: pluginName,
+    shouldAdd: true,
+  })
+
+  const { code } = babel.transform(src, {
+    plugins: [addPlugins.plugin],
+  })
+
+  return code
+}
+
+const getPluginsFromConfig = src => {
+  const getPlugins = new BabelPluginGetPluginsFromGatsbyConfig()
+
+  babel.transform(src, {
+    plugins: [getPlugins.plugin],
+  })
+
+  return getPlugins.state
+}
+
+const create = async ({ root }, { name }) => {
+  const configPath = path.join(root, `gatsby-config.js`)
+  const configSrc = await fs.readFile(configPath, `utf8`)
+
+  const prettierConfig = await prettier.resolveConfig(root)
+
+  let code = addPluginToConfig(configSrc, name)
+  code = prettier.format(code, { ...prettierConfig, parser: `babel` })
+
+  await fs.writeFile(configPath, code)
+
+  return await read({ root }, name)
+}
+
+const read = async ({ root }, id) => {
+  const configPath = path.join(root, `gatsby-config.js`)
+  const configSrc = await fs.readFile(configPath, `utf8`)
+
+  const name = getPluginsFromConfig(configSrc).find(name => name === id)
+
+  if (name) {
+    return { id, name, _message: `Installed ${id} in gatsby-config.js` }
+  } else {
+    return undefined
+  }
+}
+
+const destroy = async ({ root }, { name }) => {
+  const configPath = path.join(root, `gatsby-config.js`)
+  const configSrc = await fs.readFile(configPath, `utf8`)
+
+  const addPlugins = new BabelPluginAddPluginsToGatsbyConfig({
+    pluginOrThemeName: name,
+    shouldAdd: false,
+  })
+
+  const { code } = babel.transform(configSrc, {
+    plugins: [addPlugins.plugin],
+  })
+
+  await fs.writeFile(configPath, code)
+}
+
+class BabelPluginAddPluginsToGatsbyConfig {
+  constructor({ pluginOrThemeName, shouldAdd }) {
+    this.plugin = declare(api => {
+      api.assertVersion(7)
+
+      const { types: t } = api
+      return {
+        visitor: {
+          ExpressionStatement(path) {
+            const { node } = path
+            const { left, right } = node.expression
+
+            if (!isDefaultExport(left)) {
+              return
+            }
+
+            const plugins = right.properties.find(p => p.key.name === `plugins`)
+
+            if (shouldAdd) {
```

Comment:
One thing to note is that some plugins can be used multiple times, for instance `gatsby-source-filesystem`. This currently doesn't support it.